File: src/main/java/io/openfuture/openmessanger/web/controller/AttachmentController.java
package io.openfuture.openmessanger.web.controller;

import java.io.IOException;

import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import io.openfuture.openmessanger.service.AttachmentService;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@RequestMapping("/api/v1/attachments")
@RestController
@RequiredArgsConstructor
public class AttachmentController {

    private final AttachmentService attachmentService;

    @PostMapping
    public String handleFileUpload(final @RequestParam("file") MultipartFile file) throws IOException {
        attachmentService.upload(file);
        return file.getOriginalFilename();
    }

    @GetMapping(value = "/{fileName}", produces = MediaType.APPLICATION_OCTET_STREAM_VALUE)
    public void download(final HttpServletResponse response, @PathVariable(value = "fileName") final String fileName) throws IOException {
        final byte[] fileData = attachmentService.download(fileName);
        response.reset();
        response.setContentType(MediaType.APPLICATION_OCTET_STREAM_VALUE);
        response.setHeader(HttpHeaders.CONTENT_DISPOSITION, "attachment;filename=" + fileName);
        response.setContentLength(fileName.length());
        response.getOutputStream().write(fileData);
        response.flushBuffer();
    }

}

File: src/main/kotlin/io/openfuture/openmessanger/web/controller/AttachmentController.kt
package io.openfuture.openmessanger.web.controller

import io.openfuture.openmessanger.service.AttachmentService
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@RequestMapping("/api/v1/attachments")
@RestController
@RequiredArgsConstructor
class AttachmentController (
    val attachmentService: AttachmentService
){

    @PostMapping
    @Throws(IOException::class)
    fun handleFileUpload(@RequestParam("file") file: MultipartFile): String? {
        attachmentService.upload(file)
        return file.originalFilename
    }

    @PostMapping("/upload")
    @Throws(IOException::class)
    fun upload(@RequestParam("file") file: MultipartFile): Int {
        return attachmentService.uploadAndReturnId(file)
    }

    @GetMapping(value = ["/{fileName}"], produces = [MediaType.APPLICATION_OCTET_STREAM_VALUE])
    @Throws(IOException::class)
    fun download(response: HttpServletResponse, @PathVariable(value = "fileName") fileName: String) {
        val fileData = attachmentService.download(fileName)
        response.reset()
        response.contentType = MediaType.APPLICATION_OCTET_STREAM_VALUE
        response.setHeader(HttpHeaders.CONTENT_DISPOSITION, "attachment$fileName")
        response.setContentLength(fileData!!.size)
        response.outputStream.write(fileData)
        response.flushBuffer()
    }
}